src/reducers/user.js

```diff
@@ -1,19 +1,32 @@
 import Immutable from 'immutable';
 import get from 'lodash/get';
+import AuthenticationStates from '../enums/AuthenticationStates';
 
-const init = new Immutable.Map({authenticated: false});
+const init = new Immutable.Map({
+  authenticationState: AuthenticationStates.UNKNOWN,
+});
 
 function user(stateIn, action) {
```

Comment:
Does this do the right thing if you access Popcode while already logged in (e.g. by opening a second tab)? I suspect the current behavior will be to ask you for confirmation, which isn’t what we want.

Reply:
I understand the issue and would like to discuss before tackling it.

1. How do we convey to Firebase that the user is confirmed or not? Should we update the user data in Firebase after the user has confirmed identity? Maybe make a call similar to `saveProviderInfo()` in `firebase.js` with a boolean flag `isUserAuthenticated` on `additionalUserInfo` object.

1. When `applicationLoaded`, we can then check `isUserAuthenticated` and set the correct `authenticationState` in the store.

Reply:
@ericliu121187 we talked about this a bit on Slack but just to have it all in one place (and also I don’t think I was expressing myself very clearly):

I don’t think confirmation state is a Firebase-level concern (it’s not a persistent fact about the _user_; it’s a fact about the current user session).

I do think we want to vary the behavior on the application level based on what sequence of events yields to the user being authenticated via Firebase:

* If the user is authenticated in Firebase at application load time, just set their state directly to `AUTHENTICATED` (this should only happen if they already have a Popcode session open, or did very recently, so confirmation shouldn’t be needed)
* If the user clicks “login” and then successfully authenticates via Firebase, set the state to `PENDING_CONFIRMATION` and show the dialog
* If the user’s Firebase auth state changes to authenticated in any other scenario (meaning they logged in in a different Popcode window), set the state directly to `AUTHENTICATED`

The tricky thing about distinguishing between the latter two is that we simply [listen for changes in the Firebase auth state](https://github.com/popcodeorg/popcode/blob/3603fc88adc5b2f166832e6c5d10c11eea70c4ca/src/components/Workspace.jsx#L197-L202) and dispatch actions accordingly, totally agnostic to what initiated the auth state change.

I think probably the ideal solution would be to refactor the auth system to use [channels](https://github.com/redux-saga/redux-saga/blob/master/docs/advanced/Channels.md) to be able to consume Firebase auth changes with sagas. Then I think we could just set up a suitable system of `take`s to do the right thing in various situations.

Let me know your thoughts…!